### PR TITLE
feat(disp): add double buffered direct-mode efficient sync algorithm

### PR DESCRIPTION
**Pull request by bhspyder**
### Description of the feature or fix

Implements a more efficient double buffer sync algorithm that delays the sync of the changed areas until after the next render and only syncs the previously changed areas that did not get changed during that render. Further optimization was done to only update subregions of a previously changed area if part of it falls within an area changed by the latest render. See #4262 for more information.

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [x] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [x] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [x] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used. See a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the followings needs to be followed (see a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)): 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by bhspyder** [comment]
Not stale. Not sure if this required any changes or not, but I will resolve the conflicts on latest master so it can be merged when reviewed.

**Comment by kisvegabor** [comment]
All good from my side. I would be happy to merge it once the conflicts are resolved. 

Please pay attentions to this fix in v8: https://github.com/lvgl/lvgl/commit/28094404adab253298e0a1773b7849d6338bae15

**Comment by kisvegabor** [comment]
Not stale

**Comment by kisvegabor** [comment]
Why was it deleted?

**Comment by bhspyder** [comment]
I apologize. I was doing some cleanup on my fork and forgot that this was still outstanding on master. I will try and bring it back and reopen the pull request.

**Comment by bhspyder** [comment]
This has been caught up to the latest master. I ran code-format.py, and confirmed that the library builds and works as expected on my end so I'm not sure why a number of the checks are failing.

File: src/display/lv_display.c [hunk 1]
```diff
@@ -83,7 +83,9 @@ lv_display_t * lv_display_create(lv_coord_t hor_res, lv_coord_t ver_res)
 
     disp->inv_en_cnt = 1;
 
-    lv_display_t * disp_def_tmp = disp_def;
+    _lv_ll_init(&disp->sync_areas, sizeof(lv_area_t));
+
+    lv_disp_t * disp_def_tmp = disp_def;
```
Comment (bjsylvia) [inline src/display/lv_display.c]: replace lv_disp_t with lv_display_t

**Comment by bjsylvia** [comment]
Maybe you should replace `lv_disp_xxx` with  `lv_display_xxx` in `lv_refr.c`?

**Comment by kisvegabor** [comment]
Thank you for the update @bhspyder! 

>  Maybe you should replace lv_disp_xxx with lv_display_xxx in lv_refr.c?

Yes, it should solve it!

**Comment by bhspyder** [comment]
I was unaware of the API changes. I thought I had built it fine on my end, but I must have been on the wrong branch or project. I have been running the code-format.py script, but I always seem to have code formatting problems. I am running it windows. Not sure if that has something to do with it.

**Comment by kisvegabor** [comment]
I've also run the code formatter and now it looks good. I don't know what was the problem. Anyway, when all CI passes I will merge it.

**Comment by kisvegabor** [comment]
The failing MicroPython CI is unrelated, I'll fix it in minute.

**Pull request merged**